libraries/AP_Math/scurves.cpp

```diff
@@ -0,0 +1,645 @@
+{
+    float Beta = M_PI / tj;
+    float Alpha = Jp / 2.0;
+    float AT = Alpha * tj;
+    float VT = Alpha * (sq(tj) / 2.0 - 2.0 / sq(Beta));
+    float PT = Alpha * ((-1.0 / sq(Beta)) * tj + (1 / 6.0) * powf(tj, 3.0));
+    float A2T = Jp * tj;
+    float V2T = Jp * sq(tj);
+    float P2T = Alpha * ((-1.0 / sq(Beta)) * 2.0 * tj + (4.0 / 3.0) * powf(tj, 3.0));
+
+    enum jtype_t Jtype = jtype_t::NEGATIVE;
+    float J = Jp;
+    float T = segment[num_segs - 1].start_time + tj;
+    float A = (segment[num_segs - 1].start_accel - AT) + A2T;
+    float V = (segment[num_segs - 1].start_vel - VT) + (segment[num_segs - 1].start_accel - AT) * tj + V2T;
+    float P = (segment[num_segs - 1].start_pos - PT) + (segment[num_segs - 1].start_vel - VT) * tj + 0.5 * (segment[num_segs - 1].start_accel - AT) * sq(tj) + P2T;
+    add_segment(T, Jtype, J, A, V, P);
+}
+
+// generate three time segment raised cosine jerk profile
+void scurves::add_segments_incr_const_decr_jerk(float tj, float Jp, float Tcj)
+{
+    add_segment_incr_jerk(tj, Jp);
+    add_segment_const_jerk(Tcj, Jp);
+    add_segment_decr_jerk(tj, Jp);
+}
+
+// generate time segments for straight segment
+void scurves::add_segments_straight(float Pp)
+{
+    if (is_zero(Pp)) {
+        return;
+    }
+
+    float tj = otj;
+    float Jp = jerk_max;
+    float Ap = accel_max;
+    float Vp = vel_max;
+
+    float t2, t4, t6;
+    cal_posfast(tj, Jp, Ap, Vp, Pp / 2.0, Jp, t2, t4, t6);
+
+    add_segments_incr_const_decr_jerk(tj, Jp, t2);
+    add_segment_const_jerk(t4, 0.0);
+    add_segments_incr_const_decr_jerk(tj, -Jp, t6);
+
+    float t8 = 2.0f * (Pp / 2.0 - segment[num_segs - 1].start_pos) / segment[num_segs - 1].start_vel;
+    add_segment_const_jerk(t8, 0.0);
+
+    add_segments_incr_const_decr_jerk(tj, -Jp, t6);
+    add_segment_const_jerk(t4, 0.0);
+    add_segments_incr_const_decr_jerk(tj, Jp, t2);
+}
+
+// generate time segments to generate large curved corners
+void scurves::add_segments_curved(float Pp, float Pm)
+{
+    if (is_zero(Pp)) {
+        return;
+    }
+
+    float tj = otj;
+    float Jp = jerk_max;
+    float Ap = accel_max;
+    float Vp = vel_max;
+    float Js;
+    float Vs;
+    float Ps;
+    float tc = 0;
+    float Jc = 0;
+    float Ac = 0;
+    float Vc = 0;
+    float Pc = 0;
+
+    Pc = Pp * 2.0 / 3.0;
+    Ps = Pp - Pc;
+    Vs = MIN(Vp, MIN(MIN(safe_sqrt(Ap * Pc), powf(0.5 * Jp * sq(Pc), 1.0 / 3.0)), Pc / (2 * tj)));
+
+    float t2, t4, t6;
+    cal_posfast(tj, Jp, Ap, Vs, Ps, Js, t2, t4, t6);
+    add_segments_incr_const_decr_jerk(tj, Js, t2);
+    add_segment_const_jerk(t4, 0.0);
+    add_segments_incr_const_decr_jerk(tj, -Js, t6);
+
+    Pc = Pp - segment[num_segs - 1].start_pos;
+    Vc = segment[num_segs - 1].start_vel;
+
+    Ac = MIN(MIN(Pc / (4 * tj * tj), Vc * Vc / Pc), powf((Jp * Jp) * Pc, 1.0 / 3.0) * 6.299605249474365E-1);
+
+    Jc = powf(Ac, 3.0 / 2.0) * 1.0 / safe_sqrt(Pc) * 2.0;
+    tc = Ac / Jc;
+
+    add_segment_incr_jerk(tc, -Jc);
```

Comment:
Almost need a diagram for these!

A comment on each block of segment additions here might be nice.

Reply:
I have done a couple presentations, does that count :)